### PR TITLE
feat(layers): Port `middleware` (layers) from `ethers-rs`

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

To add examples from `ethers-rs`: `middleware` (layers)

## Solution

Implements a subset of the examples from `ethers-rs`, specifically the `NonceManagerMiddleware` and the `SignerMiddleware`.

Currently not implemented as an example, due to the implementation missing in Alloy:

- `GasEscalatorMiddleware`
- `GasOracleMiddleware`
- `PolicyMiddleware`

Some examples are not implemented due to the builder pattern being used throughout all examples in Alloy.

`CustomMiddleware` has not been implemented yet as I think it likely will require some implementation work to be able to define a robust way of handling this. Users can refer to the existing implementations as a starting point.

I've decided to refer to these as `layers` rather than `ether-rs`'s `middleware`.

### CODE REVIEW

File: examples/layers/examples/nonce_layer.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        from: Some(from),
+        value: Some(U256::from(100)),
+        to: address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into(),
+        gas_price: Some(U256::from(20e9)),
+        gas: Some(U256::from(21000)),
+        ..Default::default()
+    };
+
+    let pending = provider.send_transaction(tx.clone()).await?;
+    let tx_hash = pending.watch().await?;
```
Comment (gakonst) [inline examples/layers/examples/nonce_layer.rs:50]: The `txhash` should be available by calculating it locally, no? no need to wait an extra future for it, if you're watching we should be getting the mined transaction (or receipt?) back not the hash? this API feels weird cc @DaniPopes , maybe we already discussed

Comment (zerosnacks) [reply examples/layers/examples/nonce_layer.rs]: Fixed, now uses the local tx hash without waiting for it to be mined. `watch` sends back the tx hash whereas `get_receipt` sends back the tx receipt.

File: examples/layers/examples/nonce_layer.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        .on_client(RpcClient::new_http(http));
+
+    let tx = TransactionRequest {
+        from: Some(from),
+        value: Some(U256::from(100)),
+        to: address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into(),
+        gas_price: Some(U256::from(20e9)),
+        gas: Some(U256::from(21000)),
+        ..Default::default()
+    };
```
Comment (mattsse) [inline examples/layers/examples/nonce_layer.rs:45]: this should use the request builder API
that should work for send_transaction

File: examples/layers/examples/signer_layer.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        .on_client(RpcClient::new_http(http));
+
+    let tx = TransactionRequest {
+        nonce: Some(0),
+        value: Some(U256::from(100)),
+        to: address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into(),
+        gas_price: Some(U256::from(20e9)),
+        gas: Some(U256::from(21000)),
+        ..Default::default()
+    };
```
Comment (mattsse) [inline examples/layers/examples/signer_layer.rs:35]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
we want to showcase the request builder here

File: examples/layers/examples/signer_layer.rs (hunk 3 above)
Comment (zerosnacks) [reply examples/layers/examples/signer_layer.rs]: Updated to use the new `TransactionBuilder`, uses a `legacy` transaction.

**Comment by zerosnacks** [comment]
Merging this, requested changes were addressed

**Pull request merged**